pkg/generic/thrift/json_go116plus_amd64.go

```diff
@@ -0,0 +1,111 @@
+	"github.com/cloudwego/dynamicgo/conv/j2t"
+
+	"github.com/cloudwego/kitex/pkg/generic/descriptor"
+	"github.com/cloudwego/kitex/pkg/protocol/bthrift"
+	"github.com/cloudwego/kitex/pkg/remote/codec/perrors"
+	cthrift "github.com/cloudwego/kitex/pkg/remote/codec/thrift"
+	"github.com/cloudwego/kitex/pkg/utils"
+)
+
+// Write write json string to out thrift.TProtocol
+func (m *WriteJSON) Write(ctx context.Context, out thrift.TProtocol, msg interface{}, requestBase *Base) error {
+	if m.enableDynamicgo {
+		var msgLen int
+		_, isVoid := msg.(descriptor.Void)
+		transBuff, isString := msg.(string)
+		if !isVoid && !isString {
+			return perrors.NewProtocolErrorWithType(perrors.InvalidData, "decode msg failed, is neither void nor string")
+		} else if isVoid {
+			msgLen = 0
+		} else {
+			msgLen = len(transBuff)
+		}
+
+		if m.dty == nil {
+			return perrors.NewProtocolErrorWithMsg("svcDsc.DynamicgoDsc is nil")
+		}
+
+		// TODO: use Malloc and MallocAck to replace WriteBinary
+		buf := mcache.Malloc(msgLen + structWrapLen)
+
+		var offset int
+		offset += bthrift.Binary.WriteStructBegin(buf[offset:], m.dty.Struct().Name())
+
+		for _, field := range m.dty.Struct().Fields() {
+			if m.isClient {
+				offset += bthrift.Binary.WriteFieldBegin(buf[offset:], field.Name(), field.Type().Type().ToThriftTType(), int16(field.ID()))
+			} else {
+				// Exception field
+				if field.ID() != 0 {
+					// generic server ignore the exception, because no description for exception
+					// generic handler just return error
+					continue
+				}
+				offset += bthrift.Binary.WriteFieldBegin(buf[offset:], field.Name(), field.Type().Type().ToThriftTType(), int16(field.ID()))
+			}
+
+			if isString {
+				// encode using dynamicgo
+				cv := j2t.NewBinaryConv(m.opts)
+				v := utils.StringToSliceByte(transBuff)
+				// json []byte to thrift []byte
+				dbuf := buf[offset:offset]
+				if err := cv.DoInto(ctx, field.Type(), v, &dbuf); err != nil {
+					return err
+				} else {
+					// sometimes len of thrift []byte gets larger than len of json []byte when writing required/default fields
+					if offset+len(dbuf) > len(buf) {
+						buf = append(buf, dbuf[len(buf)-offset:]...)
+					}
+					offset += len(dbuf)
+				}
+			} else {
+				offset += bthrift.Binary.WriteStructEnd(buf[offset:])
+				offset += bthrift.Binary.WriteFieldStop(buf[offset:])
+				offset += bthrift.Binary.WriteStructEnd(buf[offset:])
+			}
+			offset += bthrift.Binary.WriteFieldEnd(buf[offset:])
+		}
+
+		// in case there is no space to write field stop
+		if offset+writeFieldStopLen > len(buf) {
+			buf = append(buf, make([]byte, (offset+writeFieldStopLen)-len(buf))...)
+		}
+		offset += bthrift.Binary.WriteFieldStop(buf[offset:])
+		bthrift.Binary.WriteStructEnd(buf[offset:])
+
+		tProt, ok := out.(*cthrift.BinaryProtocol)
+		if !ok {
+			return perrors.NewProtocolErrorWithMsg("TProtocol should be BinaryProtocol")
+		}
+		_, err := tProt.ByteBuffer().WriteBinary(buf[:offset])
```

Comment:
```golang
var buf, err = Malloc(len(b))
n = copy(buf, b)
return n, err
```
Since WriteBinary might reference the orginal buf, we can't mcache.Free buf directly, so use the above code to replace WriteBinary and then we can mcache.Free(buf) to reduce memory alloc.

Reply:
It needs once copied compared to Malloc and MallocAck, but since the latter doesn't support now, we temporarily use this way

Reply:
@jayantxie Do you mean we don't want to pass the original buf to WriteBinary of LinkBuffer?

Reply:
fixed